opm/parser/eclipse/EclipseState/Util/OrderedMap.hpp

```diff
@@ -28,124 +28,128 @@
 
 namespace Opm {
 
-template <typename T>
+template <typename KeyType, typename ValueType>
 class OrderedMap {
 private:
-    std::unordered_map<std::string , size_t> m_map;
-    std::vector<T> m_vector;
+    std::unordered_map<KeyType, size_t> m_map;
+    std::vector<std::pair<KeyType, ValueType>> m_vector;
 
 public:
-    bool hasKey(const std::string& key) const {
-        auto iter = m_map.find(key);
-        if (iter == m_map.end())
-            return false;
-        else
-            return true;
+
+    std::size_t count(const KeyType& key) const {
+        return this->m_map.count(key);
+    }
+
+
+
+    ValueType& operator[](const KeyType& key) {
+        if (this->count(key) == 0)
+            this->insert( std::make_pair(key, ValueType()));
+
+        return this->at(key);
     }
 
 
-    void insert(std::string key, T value) {
-        if (hasKey(key)) {
-            auto iter = m_map.find( key );
+    void insert(std::pair<KeyType,ValueType> pair) {
+        if (this->count(pair.first) > 0) {
+            auto iter = m_map.find( pair.first );
             size_t index = iter->second;
-            m_vector[index] = value;
+            m_vector[index] = pair;
         } else {
             size_t index = m_vector.size();
-            m_vector.push_back( std::move( value ) );
-            m_map.insert( std::pair<std::string, size_t>(key , index));
+            m_map.insert( std::pair<std::string, size_t>(pair.first , index));
```

Comment:
This will copy `pair.first` twice.  The value type of a `std::unordered_map<K,V>` is `std::pair<const K, V>`, not `std::pair<K,V>`. In other words, here you're creating a pair that must be copied/converted to another pair before being inserted into the `m_map` data member.

Reply:
OK - thanks; I'll fix it here: #670

Reply:
I think `m_map.emplace(pair.first, index)` would be simplest, most readable and most performant. But using `pair` as a variable name is unfortunate...

Reply:
Yes - I agree about `emplace` - initially that did not compile; but that was due to something else. Using `emplace` and also renamed variable `pair` &rightarrow; `key_value_pair` here: #670